antelope_contracts/contracts/erc20/include/erc20/erc20.hpp

```diff
@@ -11,28 +11,100 @@ using namespace intx;
 
 namespace erc20 {
 
+checksum256 make_key(const uint8_t* ptr, size_t len) {
+    uint8_t buffer[32]={0};
+    check(len <= sizeof(buffer), "invalida size");
+    memcpy(buffer, ptr, len);
+    return checksum256(buffer);
+}
+
+checksum256 make_key(bytes data){
+    return make_key((const uint8_t*)data.data(), data.size());
+}
+
 class [[eosio::contract]] erc20 : public contract {
    public:
     using contract::contract;
 
+   struct bridge_message_v0 {
+      eosio::name        receiver;
+      bytes              sender;
+      eosio::time_point  timestamp;
+      bytes              value;
+      bytes              data;
+
+      EOSLIB_SERIALIZE(bridge_message_v0, (receiver)(sender)(timestamp)(value)(data));
+   };
+
+   using bridge_message_t = std::variant<bridge_message_v0>;
+
     [[eosio::on_notify("*::transfer")]] void transfer(eosio::name from, eosio::name to, eosio::asset quantity, std::string memo);
 
     // evm runtime will call this to notify erc20 about the message from 'from' with 'data'.
-    [[eosio::action]] void onbridgemsg(name receiver, const bytes& sender, const time_point& timestamp, const bytes& value, const bytes& data);
-    [[eosio::action]] void init();
+    [[eosio::action]] void onbridgemsg(const bridge_message_t &message);
+    [[eosio::action]] void init(uint64_t nonce);
+
+    [[eosio::action]] void regtoken(uint64_t nonce, eosio::name eos_contract_name, 
```

Comment:
The contract should get the nonce from the EOS EVM Contract table rather than expect the user to provide the nonce value. This applies to the `init` action as well, which I suggest we rename to `upgrade`.

Reply:
fixed